JSDoc: can't resolve import(...) if it's defined at the bottom of a file

**TypeScript Version:** 2.9.0-dev.20180418

**Search Terms:**
 - JSDoc
 - import
 - types

**Code**

```ts
// @filename: interfaces.d.ts
export interface Bar {
    prop: string
}

// @filename: usage.js
/** @type {Bar} */
export let bar;

// TS raises following error:
// TS2307: Cannot find module './interfaces'
/** @typedef {import('./interfaces').Bar} Bar */
```

**Expected behavior:**
TS successfully resolves requested module. No errors should be thrown.

**Actual behavior:**
TS can't find a requested module if `import` is used at the bottom of a file.

--- a/src/compiler/program.ts
+++ b/src/compiler/program.ts
@@ -1624,6 +1624,9 @@ namespace ts {
                     collectDynamicImportOrRequireCalls(node);
                 }
             }
+            if ((file.flags & NodeFlags.PossiblyContainsDynamicImport) || isJavaScriptFile) {
+                collectDynamicImportOrRequireCalls(file.endOfFileToken);
+            }
 
             file.imports = imports || emptyArray;
             file.moduleAugmentations = moduleAugmentations || emptyArray;

--- a/src/compiler/program.ts
+++ b/src/compiler/program.ts
@@ -2007,7 +2007,8 @@ namespace ts {
                         && !options.noResolve
                         && i < file.imports.length
                         && !elideImport
-                        && !(isJsFile && !options.allowJs);
+                        && !(isJsFile && !options.allowJs)
+                        && (isInJavaScriptFile(file.imports[i]) || !(file.imports[i].flags & NodeFlags.JSDoc));
 
                     if (elideImport) {
                         modulesWithElidedImports.set(file.path, true);

--- a/tests/baselines/reference/importTypeResolutionJSDocEOF.symbols
+++ b/tests/baselines/reference/importTypeResolutionJSDocEOF.symbols
@@ -1,0 +1,14 @@
+=== tests/cases/compiler/interfaces.d.ts ===
+export interface Bar {
+>Bar : Symbol(Bar, Decl(interfaces.d.ts, 0, 0))
+
+    prop: string
+>prop : Symbol(Bar.prop, Decl(interfaces.d.ts, 0, 22))
+}
+
+=== tests/cases/compiler/usage.js ===
+/** @type {Bar} */
+export let bar;
+>bar : Symbol(bar, Decl(usage.js, 1, 10))
+
+/** @typedef {import('./interfaces').Bar} Bar */

--- a/tests/baselines/reference/importTypeResolutionJSDocEOF.types
+++ b/tests/baselines/reference/importTypeResolutionJSDocEOF.types
@@ -1,0 +1,14 @@
+=== tests/cases/compiler/interfaces.d.ts ===
+export interface Bar {
+>Bar : Bar
+
+    prop: string
+>prop : string
+}
+
+=== tests/cases/compiler/usage.js ===
+/** @type {Bar} */
+export let bar;
+>bar : Bar
+
+/** @typedef {import('./interfaces').Bar} Bar */

--- a/tests/cases/compiler/importTypeResolutionJSDocEOF.ts
+++ b/tests/cases/compiler/importTypeResolutionJSDocEOF.ts
@@ -1,0 +1,13 @@
+// @allowJs: true
+// @noEmit: true
+// @checkJs: true
+// @filename: interfaces.d.ts
+export interface Bar {
+    prop: string
+}
+
+// @filename: usage.js
+/** @type {Bar} */
+export let bar;
+
+/** @typedef {import('./interfaces').Bar} Bar */

